### PR TITLE
TravisCI: Check for PEP8

### PR DESCRIPTION
**Pull request by tammoippen**
This PR lets TravisCI check changed Python files for PEP8 compliance, as requested by #188. The file `extras/check_code_style.sh` now also checks for PEP8.

### CODE REVIEW

**Comment by heplesser** [comment]
:+1: from my side. @apeyser, would you take a look, too?

File: build.sh [hunk 1]
```diff
@@ -111,6 +111,9 @@ else
 fi
 format_error_files=""
 
+# Ignore those PEP8 rules
+PEP8_IGNORES="E121,E123,E126,E226,E24,E704"
```
Comment (apeyser) [inline build.sh]: Should the E24 be E24x? I don't see an E24 in the pep8 script's documentation.
Where did this particular choice of pep8 exceptions come from?

File: build.sh [hunk 2]
```diff
@@ -148,8 +148,20 @@ for f in $file_names; do
       fi
 
       ;;
+    *.py )
+      set +e # pep8 returns 1 if the file does not comply with PEP8
+      echo "Check PEP8 on file $f:"
+      pep8_result=`pep8 --first $f`
```
Comment (apeyser) [inline build.sh]: Why not just
````
if ! pep8_result=`pep8 --first $f`; then
````

File: extras/parse_travis_log.py [hunk 3]
```diff
@@ -177,6 +180,35 @@ def process_static_analysis(f, line):
...
 
+def process_pep8(f, line):
+    """
+    Process PEP8 output for a certain file.
+    """
+    # Check PEP8 on file pynest/nest/lib/hl_api_connections.py:
+    filename = line.split(' ')[-1].strip()[0:-1]
+    d = set()
+    res = {filename: d}
+    while True:
```
Comment (apeyser) [inline extras/parse_travis_log.py]: Options here: 
````
fiter = iter(f)
for line in fiter:
```
and use the same iter at 206
or ... comment what's going on here with walking through the file.

File: build.sh (hunk 1 above)
Comment (tammoippen) [inline build.sh]: Hi @apeyser,

Without it TravisCI complains, and my local machine not: `extras/parse_travis_log.py:285:45: E126 continuation line over-indented for hanging indent`
The line looks much nicer with this over-indention. On my Mac when I execute `pep8 --help` I get:
```
pep8 --help
Usage: pep8 [options] input ...

Options:
...
  --ignore=errors      skip errors and warnings (e.g. E4,W) (default:
                       E121,E123,E126,E226,E24,E704)
  --show-source        show source code for each error
```
i.e. `E126` is ignored here by default. Maybe we should only ignore `E126`? Also see [here](https://github.com/PyCQA/pycodestyle/issues/167) for discussion.

Comment (apeyser) [inline build.sh]: My python3?-pep8 doesn't give me the defaults. Maybe I should pull up the script and see what the defaults actually are.

File: extras/check_code_style.sh [hunk 4]
```diff
@@ -216,8 +228,20 @@ for f in $file_names; do
       $CPPCHECK --enable=all --inconclusive --std=c++03 $f
 
       ;;
+    *.py )
+      set +e # pep8 returns 1 if the file does not comply with PEP8
+      echo "Check PEP8 on file $f:"
+      pep8_result=`pep8 --first $f`
+
+      if [ "$?" -gt "0" ]; then
```
Comment (apeyser) [inline extras/check_code_style.sh]: I thought I commented on this, but can't find it, so...
Why not:
````
if ! pep8_result=`pep8 --first $f`; then
```

File: extras/check_code_style.sh [hunk 5]
```diff
@@ -216,8 +229,20 @@ for f in $file_names; do
       $CPPCHECK --enable=all --inconclusive --std=c++03 $f
 
       ;;
+    *.py )
+      set +e # pep8 returns 1 if the file does not comply with PEP8
+      echo "Check PEP8 on file $f:"
+      pep8_result=`pep8 --first --ignore=$PEP8_IGNORES $f`
+
+      if [ "$?" -gt "0" ]; then
```
Comment (apeyser) [inline extras/check_code_style.sh]: Why not: if ! pep8_result=`pep8 ...` ?

File: extras/parse_travis_log.py (hunk 3 above)
Comment (tammoippen) [inline extras/parse_travis_log.py]: I go through the file only once and `f` stores the "pointer" into where we are at reading the file. When i use `iter(f)`, each function has its own state there, correct? Then I have to rewrite the script, which might be not a bad idea, as it will make the individual parsing functions simpler. But I think this would be a topic for another PR.

Comment (tammoippen) [inline extras/parse_travis_log.py]: I go through the file only once and `f` stores the "pointer" into where we are at reading the file. When i use `iter(f)`, each function has its own state there, correct? Then I have to rewrite the script, which might be not a bad idea, as it will make the individual parsing functions simpler. But I think this would be a topic for another PR.

Comment (apeyser) [inline extras/parse_travis_log.py]: In fact, no -- the underlying file handles moves forward, so:
````
f = open(file, 'r')
f1 = iter(f)
f2 = iter(f)
for l in f1: print l
````

will move f2 forward as well as f1. But in general for iterators, you would be right. And you're probably right that everything should just move the iterator instead of the file handle around... but it is too wide for this pr.

I'd just do:
````
for line in f: ....
````

since that's anyway the same thing.

Comment (tammoippen) [inline extras/parse_travis_log.py]: There is another nested loop continuing the iteration throught the file:

```Python
if line.startswith('+echo ' + filename):
    while True:
        line = f.readline()
```

Does it work there, too?

Comment (apeyser) [inline extras/parse_travis_log.py]: Should, since the iterator is simply an adaptor on the common underlying buffer and file descriptor at the very bottom.

File: extras/check_code_style.sh [hunk 6]
```diff
@@ -229,7 +251,10 @@ if [ "$format_error_files" != "" ]; then
     echo "  $f"
   done
   echo
-  echo "Perform '$CLANG_FORMAT -i <filename>' on them, otherwise TravisCI will report failure!"
+  echo "TravisCI will report failure!"
+  echo "- On C/CPP files, perform '$CLANG_FORMAT -i <filename>' on them."
+  echo "- On Python files, perform 'pep8ify -w <filename>' on them and"
```
Comment (jougs) [inline extras/check_code_style.sh]: Please remove the suffix "on them". See [here](http://uncyclopedia.wikia.com/wiki/Redundancy).

File: extras/parse_travis_log.py [hunk 7]
```diff
@@ -63,18 +64,20 @@ def process_installcheck(f):
 
 def process_changed_files(f):
     """
-    Run through the file f until a line starts with '+file_names='. After the
-    equal sign is a space separated list of changed files in the PR. Returns that
-    list.
+    Run through the file f until a line starts with '+file_names='.
+    After the equal sign is a space separated list of changed files in the PR.
```
Comment (jougs) [inline extras/parse_travis_log.py]: The equal sign is followed by a space separated list of changed files in the PR.

File: extras/parse_travis_log.py (hunk 3 above)
Comment (tammoippen) [inline extras/parse_travis_log.py]: @apeyser I structured the script with the `readline()` method. When I replace it with the iterator, I get errors like:

    ValueError: Mixing iteration and read methods would lose data

As this is not directly related to PEP8, I would like to put this into another issue, and here just document, what the lines are intended to do.

Comment (apeyser) [inline extras/parse_travis_log.py]: @tammoippen 

Sounds reasonable. I see in the documentation that the read-ahead buffer is broken, that it's only used in next calls but not readline calls.  It's even built on top of a FILE*, so that you have buffers of buffers on top of the POSIX buffers on top of the hardware buffers...
(This is also the answer to why I'm suspicious of all directives coming out of that design camp)

**Comment by apeyser** [comment]
:+1:
